test/bn_3d_peract_test.cpp

```diff
@@ -108,11 +108,11 @@ struct verify_forward_train_3d_bn_per_activation
             runMean = tensor<U>{rs_n_batch, rs_channels, rs_depth, rs_height, rs_width};
             runVar  = tensor<U>{rs_n_batch, rs_channels, rs_depth, rs_height, rs_width};
 
-            const U Data_scale = static_cast<U>(0.001);
+            const double Data_scale = 0.001;
```

Comment:
why is this `double` and not `U`? Just wondering.

Reply:
Potentially computing in a higher precision domain may give slightly better distribution. It may not be so critical since distribution calculation itself heavily depends on rounding and other factors.
for example distribution for range 1000 and scale 0.001 casted to float at the end (double case, blue) or during calculations (U-case, orange):
![chart](https://github.com/ROCmSoftwarePlatform/MIOpen/assets/489073/789e61bb-978a-41b5-9005-6813944e470c)

Another point - since x86 does not support `half` (and other low-precision ML types) natively, all the calculations in that domain may not be so efficient and thus it's much faster to calculate in a native domain and convert the result at the end.

Reply:
Makes sense. Although, I don't understand the plot. What am I looking at?

Reply:
It's a probability density plot.
We generate "discreet" random values in range [-0.999; 0.999] and count how many numbers come into the bins [-0.999; 0.998), [0.998; 0.997) ... [0; 0.001) ... [0.999; 1). So for the uniform distribution it must be linear.
By "discreet" I mean we generate true discreet values in range  [-999; 999] and multiply them by 0.001 so mathematically they are discreet with 0.001 step, but due to finite precision of fp values they may have a bit different nature and thus we can get some spikes because rounding and loss of precision shifted some values to the different bins.

Reply:
OK, I see. It is strange how you see variation in bucket density only on left side. I would imagine this would happen throughout.

Reply:
That strange behavior is not so critical for the "real" numbers. For example the sequence 0.9936, 0.9944, 9.9956, 0.99964, 0.99976 generates distribution like 0, 2, 0, 1. It's just not so perfectly numerically uniform. It is still discreet - we don't have any values in the middle, it still uniform since each distinctive value has uniform probability, but numerically the difference between two neighbor values is fluctuating.